src/SAWScript/SolverCache.hs

```diff
@@ -0,0 +1,632 @@
+which use 'applyProverToGoal', the type of a result is:
+@Maybe [(ExtCns Term, FirstOrderValue)]@, where 'Nothing' represents a result
+of "unsat," and 'Just' represents a result of "sat" along with a list of
+counterexamples. Since 'ExtCns' contains execution-specific 'VarIndex'es, we
+don't want to save these results directly. Instead, we represent each 'ExtCns'
+as its index in the 'satVariables' field of 'SATQuery' (which is where they
+were obtained by the backends in the first place).
+-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE ViewPatterns #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module SAWScript.SolverCache
+  ( SolverBackend(..)
+  , allBackends
+  , sbvBackends
+  , SolverBackendOption(..)
+  , optionBackends
+  , SolverBackendVersions
+  , SolverCacheKey(..)
+  , mkSolverCacheKey
+  , SolverCacheValue(..)
+  , toSolverCacheValue
+  , fromSolverCacheValue
+  , SolverCache(..)
+  , lazyOpenSolverCache
+  , openSolverCache
+  , SolverCacheOp
+  , lookupInSolverCache
+  , insertInSolverCache
+  , setSolverCachePath
+  , printSolverCacheByHex
+  , cleanSolverCache
+  , printSolverCacheStats
+  , testSolverCacheStats
+  ) where
+
+import System.Directory (createDirectoryIfMissing)
+import Control.Exception (try, SomeException)
+import Control.Monad (when, forM_)
+import System.Timeout (timeout)
+
+import GHC.Generics (Generic)
+import Data.IORef (IORef, newIORef, modifyIORef, readIORef)
+import Data.Time.Clock (UTCTime, getCurrentTime)
+import Data.Tuple.Extra (first, firstM)
+import Data.List (elemIndex, intercalate, isPrefixOf)
+import Data.Maybe (fromMaybe, isJust)
+import Data.Functor ((<&>))
+import Numeric (readHex)
+
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+
+import qualified Data.Text as T
+import Data.Text.Encoding (encodeUtf8)
+import Text.Printf (printf)
+
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+
+import qualified Crypto.Hash.SHA256 as SHA256
+
+import Data.Aeson ( FromJSON(..), ToJSON(..), FromJSONKey(..), ToJSONKey(..)
+                  , (.:), (.:?), (.=), fromJSON )
+import qualified Data.Aeson as JSON
+import Codec.CBOR.JSON (encodeValue, decodeValue)
+import Codec.Serialise (Serialise(..))
+
+import qualified Database.LMDB.Simple as LMDB
+import qualified Database.LMDB.Simple.Extra as LMDB
+
+import qualified Data.SBV.Dynamic as SBV
+
+import Verifier.SAW.FiniteValue
+import Verifier.SAW.SATQuery
+import Verifier.SAW.ExternalFormat
+import Verifier.SAW.SharedTerm
+
+import SAWScript.Options
+import SAWScript.Proof
+
+
+-- Orphan Instances and Helper Functions ---------------------------------------
+
+deriving instance Generic FirstOrderType
+deriving instance Generic FirstOrderValue
```

Comment:
Do these instances need to be orphans, or could they instead be moved to `Verifier.SAW.FiniteValue`, where these data types are defined? I'm hoping it's the latter, since that would avoid needing to use `-fno-warn-orphans`.

Reply:
Does your suggestion include the `ToJSON`/`FromJSON` instances as well? If not, `-fno-warn-orphans` would still need to be used for this file.

Reply:
Yes, the `{To,From}JSON` instances would also need to be moved to avoid them being orphans.